xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+
+    DestinationPortMatcher(int port) {
+      this.port = port;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return port == args.getDestinationPort();
+    }
+  }
+
+  private static final class EvaluateArgs {
+    private final Metadata metadata;
+    private final ServerCall<?,?> serverCall;
+    // https://github.com/envoyproxy/envoy/blob/63619d578e1abe0c1725ea28ba02f361466662e1/api/envoy/config/rbac/v3/rbac.proto#L238-L240
+    private static final int URI_SAN = 6;
+    private static final int DNS_SAN = 2;
+
+    private EvaluateArgs(Metadata metadata, ServerCall<?,?> serverCall) {
+      this.metadata = metadata;
+      this.serverCall = serverCall;
+    }
+
+    private String getPath() {
+      return serverCall.getMethodDescriptor().getFullMethodName();
+    }
+
+    private Collection<String> getPrincipalNames() {
+      SSLSession sslSession = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_SSL_SESSION);
+      Set<String> principalNames = new HashSet<>();
+      try {
+        Certificate[] certs = sslSession == null ? null : sslSession.getPeerCertificates();
+        if (certs == null || certs.length < 1) {
+          return principalNames;
+        }
+        X509Certificate cert = (X509Certificate)certs[0];
+        if (cert == null || cert.getSubjectDN() == null) {
+          return principalNames;
+        }
+        Collection<List<?>> names = cert.getSubjectAlternativeNames();
+        if (names != null) {
+          for (List<?> name : names) {
+            if (URI_SAN == (Integer) name.get(0)) {
+              principalNames.add((String) name.get(1));
+            }
+          }
+          if (!principalNames.isEmpty()) {
+            return principalNames;
+          }
+          for (List<?> name : names) {
+            if (DNS_SAN == (Integer) name.get(0)) {
+              principalNames.add((String) name.get(1));
+            }
+          }
+          if (!principalNames.isEmpty()) {
+            return principalNames;
+          }
+        }
+        return Collections.singleton(cert.getSubjectDN().getName());
+      } catch (SSLPeerUnverifiedException | CertificateParsingException ex) {
+        return null;
+      }
+    }
+
+    private Metadata getHeader() {
+      return metadata;
+    }
+
+    private InetAddress getDestinationIp() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_LOCAL_ADDR);
+      return addr == null ? null : ((InetSocketAddress) addr).getAddress();
+    }
+
+    private InetAddress getSourceIp() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+      return addr == null ? null : ((InetSocketAddress) addr).getAddress();
+    }
+
+    private int getDestinationPort() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_LOCAL_ADDR);
+      return addr == null ? -1 : ((InetSocketAddress) addr).getPort();
+    }
+  }
+
+  abstract static class Matcher {
```

Comment:
Looks `interface` is more suitable? `Matcher` itself doesn't define any blueprint.

Reply:
Right.